[replicate] support simpler device_id

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100217
* #100216
* #100131

Allow passing in `device_id=[device]` regardless of CPU or GPU. We
modify the kwarg as needed to pass to DDP.

rohan-varma committed:
[replicate] support simpler device_id

Allow passing in `device_id=[device]` regardless of CPU or GPU. We
modify the kwarg as needed to pass to DDP.

[ghstack-poisoned]

awgu commented on test/distributed/_composable/test_replicate.py:
```diff
@@ -213,6 +214,49 @@ def test_replicate_with_kwargs(self):
         )
         self._compare_module(model, replicate_model)
 
+    def test_replicate_device_id(self):
```
```
AssertionError: Torch not compiled with CUDA enabled
```
```suggestion
    @skip_if_lt_x_gpu(2)
    def test_replicate_device_id(self):
```

awgu commented on torch/distributed/_composable/replicate.py:
```diff
@@ -1,4 +1,5 @@
 from typing import Any, Dict, Iterable, List, Optional, Set, Tuple
```
nit: Run lint for this file.

awgu reviewed (approved):
Stamp to unblock, but this needs two minor changes (lint and test decorator)

rohan-varma committed:
Update on "[replicate] support simpler device_id"


Allow passing in `device_id=[device]` regardless of CPU or GPU. We
modify the kwarg as needed to pass to DDP.

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
